pandas/tests/test_strings.py

```diff
@@ -34,12 +186,82 @@ def test_api(self):
         assert isinstance(Series(['']).str, strings.StringMethods)
 
-        # GH 9184
-        invalid = Series([1])
-        with tm.assert_raises_regex(AttributeError,
-                                    "only use .str accessor"):
-            invalid.str
-        assert not hasattr(invalid, 'str')
+    @pytest.mark.parametrize('dtype', [object, 'category'])
+    @pytest.mark.parametrize('box', [Series, Index])
+    def test_api_per_dtype(self, box, dtype, all_skipna_inferred_dtypes):
+        # one instance of parametrized fixture
+        inferred_dtype, values = all_skipna_inferred_dtypes
+
+        t = box(values, dtype=dtype)  # explicit dtype to avoid casting
+
+        # TODO: get rid of these xfails
+        if dtype == 'category' and inferred_dtype in ['period', 'interval']:
+            pytest.xfail(reason='Conversion to numpy array fails because '
+                         'the ._values-attribute is not a numpy array for '
+                         'PeriodIndex/IntervalIndex; see GH 23553')
+
+        types_passing_constructor = ['string', 'unicode', 'empty',
+                                     'bytes', 'mixed', 'mixed-integer']
+        if inferred_dtype in types_passing_constructor:
+            # GH 6106
+            assert isinstance(t.str, strings.StringMethods)
+        else:
+            # GH 9184, GH 23011, GH 23163
+            with tm.assert_raises_regex(AttributeError, 'Can only use .str '
+                                        'accessor with string values!'):
+                t.str
+            assert not hasattr(t, 'str')
+
+    def test_api_mi_raises(self):
+        mi = MultiIndex.from_arrays([['a', 'b', 'c']])
+        with tm.assert_raises_regex(AttributeError, 'Can only use .str '
+                                    'accessor with Index, not MultiIndex'):
+            mi.str
+        assert not hasattr(mi, 'str')
+
+    @pytest.mark.parametrize('dtype', [object, 'category'])
+    @pytest.mark.parametrize('box', [Series, Index])
+    def test_api_per_method(self, box, dtype,
+                            all_allowed_skipna_inferred_dtypes,
+                            all_string_methods):
+        # this test does not check correctness of the different methods,
+        # just that the methods work on the specified (inferred) dtypes,
+        # and raise on all others
+
+        # one instance of each parametrized fixture
+        inferred_dtype, values = all_allowed_skipna_inferred_dtypes
+        method_name, minimal_args = all_string_methods
+
+        # TODO: get rid of these xfails
+        if (method_name in ['partition', 'rpartition'] and box == Index
+                and inferred_dtype != 'bytes'):
+            pytest.xfail(reason='Method not nan-safe on Index; see GH 23558')
+        if (method_name == 'get_dummies' and box == Index
+                and inferred_dtype == 'empty' and dtype == object):
+            pytest.xfail(reason='Need to fortify get_dummies corner case')
```

Comment:
This case wasn't an issue up until now, because `Index.str` raised if the inferred dtype was `'empty'`. As such, I haven't opened an issue for it yet.